### DISCUSSION TITLE
SDF failed to open file

### DISCUSSION THREAD

**Discussion by Vlasovian** [epochpic/epoch#205] [General]
Hi team,

When opening a large number of files in python, the SDF reader fails usually after a few thousand files. The error message reads:

Reading file Data240421/021019.sdf
Traceback (most recent call last):

  File "/home/physics/phrxss/epoch/epoch1d/B_Spectra.py", line 131, in <module>
    data = sh.getdata(j, 'Data240421')

  File "/home/physics/phrxss/.local/lib/python3.7/site-packages/sdf_helper/sdf_helper.py", line 1705, in getdata

OSError: Failed to open file: 'Data240421/021019.sdf'

I should note that the file is not missing, and does open when reading a smaller number of files, and that the sdf_helper.py file has been modified to account for the change to n_zeros.

Thanks in advance

**Comment by TomGoffrey** [comment]
My first guess was that you're hitting the limit for the number of open files, but I think that should raise an IOError rather than OSError. 

Is that the complete error you message you receive? Is it always the same file that fails/same number of files before an error?

It's probably also worth checking you can reproduce the error with an unmodified version of sdfhelper by passing the filename directly, rather than the integer.

**Reply by Vlasovian** [reply]
The number of files read before it fails is usually just over 4000, although it has failed after 1000ish too (using sh.getdata(j, 'Data240421'), and a modified sdf_helper). I also got the following error messages:

The history saving thread hit an unexpected error (OperationalError('unable to open database file')).History will not be written to the database.This command failed to be executed because an error occurred while trying to get the file code from Spyder's editor. The error was:

Exception in thread IPythonHistorySavingThread:

and

OSError: [Errno 24] Too many open files: '/home/physics/phrxss/epoch/epoch1d/B_Spectra.py'

Passing through the full filename, i.e., something like sh.getdata("Data240421/" + "%0.6d" % (j , )), and using an unmodified sdf_helper, the reader alternates between failing at 000010.sdf and 004009.sdf.

**Reply by TomGoffrey** [reply]
In that case, I think my initial suspicion was correct, and you're running into issues with too many open files.

I'm guessing you're doing some FFTs, and loading all the data into a large array? If so, I've done something like this in the past:

```python
# flist is a list of filenames
def get_time_series(flist, varname):
    var_array = []
    for f in flist:
        data = sh.getdata(f, verbose=False)
        var_array.append(np.copy(getattr(data, varname).data))
    return np.array(var_array)
```

Disclaimer: Not actual code I've run - might not work, but I think this should avoid the issue you're seeing.

**Reply by Vlasovian** [reply]
So I've given something along these lines a go, but it doesn't seem to have worked. I think I can get away with sampling the timesteps less frequently for now, though. Thanks.

**Reply by keithbennett** [reply]
It depends what you mean by "something along these lines". I think that the root of the problem is that the file handle is not closed until the structure returned by "getdata" is either explicitly destroyed or goes out of scope and gets garbage-collected. In the method posted by @TomGoffrey, the contents of the "data" structure is copied and then goes out of scope at the end of the "for" block, so the structure will get garbage-collected and at that point the file will be closed. This could be made more explicit by adding "del data" at the end of the "for" block. However, if any reference to the data structure is maintained then the file will not be closed and you will eventually run out of filehandles.
